stdlib/asm/crypto/dsa/falcon.masm

```diff
@@ -0,0 +1,277 @@
+    push.0.0.0.0
+
+    repeat.128
+        dup.4
+        mem_loadw
+
+        exec.normalize_word
+
+        dup.5
+        mem_storew
+
+        movup.5
+        add.1
+        movdn.5
+
+        movup.4
+        add.1
+        movdn.4
+    end
+
+    dropw
+    drop
+    drop
+end
+
+# Given four elements on stack top, this routine computes squared norm of that
+# vector ( read polynomial ) with four coefficients.
+#
+# Imagine, given vector is f, which is described as
+#
+# f = [a0, a1, a2, a3]
+#
+# Norm of that vector is
+#
+# √(a0 ^ 2 + a1 ^ 2 + a2 ^ 2 + a3 ^ 2)
+#
+# But we need squared norm, which is just skipping the final square root operation.
+#
+# Expected stack state :
+#
+# [a0, a1, a2, a3, ...]
+#
+# Final stack state :
+#
+# [b, ...] | b = a0 ^ 2 + a1 ^ 2 + a2 ^ 2 + a3 ^ 2
+proc.squared_norm_word
+    dup
+    mul
+
+    swap
+    dup
+    mul
+
+    add
+
+    swap
+    dup
+    mul
+
+    add
+
+    swap
+    dup
+    mul
+
+    add
+end
+
+# Given a degree 512 polynomial in coefficient form, as starting (absolute) memory address 
+# on stack, this routine computes squared norm of that vector, using following formula
+#
+# Say, f = [a0, a1, a2, ..., a510, a511]
+#      g = sq_norm(f) = a0 ^ 2 + a1 ^ 2 + ... + a510 ^ 2 + a511 ^ 2
+#
+# Expected input stack state :
+#
+# [f_start_addr, ...] | f_addr`i` holds f[(i << 2) .. ((i+1) << 2)]
+#
+# Consecutive 127 addresses on stack can be computed using `INCR` instruction, because memory 
+# addresses are consecutive i.e. monotonically increasing by 1.
+#
+# Final stack state :
+#
+# [g, ...] | g = sq_norm(f)
+export.squared_norm_poly512
+    push.0.0.0.0.0
+
+    repeat.128
+        dup.5
+        mem_loadw
+
+        exec.squared_norm_word
+        add
+
+        swap
+        add.1
+        swap
+
+        push.0.0.0.0
+    end
+
+    dropw
+    swap
+    drop
+end
+
+# Falcon-512 Digital Signature Verification routine
+#
+# Given four degree-512 polynomials, using absolute memory addresses on stack ( i.e. 
+# 512 memory addresses will be placed on stack ), this routine checks whether it's a valid
+# Falcon signature or not.
+#
+# Four degree-512 polynomials, which are provided ( in order )
+#
+# f = [f0, f1, ..., f510, f511] -> decompressed Falcon-512 signature
+# g = [g0, g1, ..., g510, g511] -> public key used for signing input message
+# h = [h0, h1, ..., h510, h511] -> input message hashed using SHAKE256 XOF and converted to polynomial
+# k = [k0, k1, ..., k510, k511] -> [abs(i) for i in f] | abs(a) = a < 0 ? 0 - a : a
+#
+# Each of these polynomials are represented using 128 absolute memory addresses, each holding
+# four consecutive coefficients such as
+#
+# f`i` holds f[(i << 2) .. ((i+1) << 2)] | i ∈ [0..128)
+#
+# Expected stack state :
+#
+# [f_addr0, f_addr1, ..., f_addr126, f_addr127, g_addr0, g_addr1, ..., g_addr126, g_addr127, 
+#  h_addr0, h_addr1, ..., h_addr126, h_addr127, k_addr0, k_addr1, ..., k_addr126, k_addr127, ...] | 512 absolute memory addresses
```

Comment:
Is this comment still accurate? I thought we are providing just starting addresses via the stack (e.g., `f_start_addr`) rather than consecutive addresses.

Reply:
Thanks for pointing this out @bobbinth , I've corrected this in 12cfe44ef2428a192ef089e83f8303e36dec9ca9.